### PR TITLE
Switch to REST API v2

### LINKED ISSUE #43: Listing contexts fails when managing a large number of contexts
With growing number of contexts and variables in CircleCi, we are getting more often this error:
`Error: failed to load context list: Non-nullable field was null.
Something unexpected happened.`

Set `-parallelism=1` for terraform plan is helped for some time, but after add more new contexts, we started to see this error again.

Also, after add 18 contexts via terraform, we got email from CircleCi support with request to explain why we produce so many requests  to CircleCi v1 API and have recommended to switch to v2 REST API:
https://circleci.com/docs/api/v2/#tag/Context

### PR DESCRIPTION
**Pull request by bendrucker**
Adopts https://circleci.com/docs/api/v2/ via https://github.com/CircleCI-Public/circleci-cli/tree/master/api

This will involve:

* Replacing the GraphQL client with the context rest client supplied by the `api` package
* Discarding the forked/unmaintained community clients in favor of the generic [rest](https://github.com/CircleCI-Public/circleci-cli/tree/master/api/rest) client
  * Implement some client methods here to make the appropriate v2 API calls for project environment variables

This change should be non-breaking for most provider users.

Opening this as a draft while I work through it.

Closes #43

### CODE REVIEW

**Comment by bendrucker** [comment]
Ok, some progress!

```
make testacc TESTARGS='-run ^TestAccCircleCIContextDataSource$'
TF_ACC=1 go test ./... -v -run ^TestAccCircleCIContextDataSource -timeout 120m
?   	github.com/mrolla/terraform-provider-circleci	[no test files]
=== RUN   TestAccCircleCIContextDataSource
--- PASS: TestAccCircleCIContextDataSource (1.69s)
PASS
ok  	github.com/mrolla/terraform-provider-circleci/circleci	1.944s
?   	github.com/mrolla/terraform-provider-circleci/circleci/client	[no test files]
?   	github.com/mrolla/terraform-provider-circleci/circleci/client/rest	[no test files]
```

I had to inline the REST client in order to make some fixes. The API has some pretty rough edges and getting things working predictably, especially when it comes to error handling, will be a bit of a chore. But overall a lot of improvement here from the provider end of things.

**Comment by bendrucker** [comment]
After making some fixes against a copy of the API client, all's ✅ with contexts!

```
TF_ACC=1 go test ./... -v -run TestAccCircleCIContext -timeout 120m
?   	github.com/mrolla/terraform-provider-circleci	[no test files]
=== RUN   TestAccCircleCIContextDataSource
--- PASS: TestAccCircleCIContextDataSource (1.85s)
=== RUN   TestAccCircleCIContextEnvironmentVariable_basic
--- PASS: TestAccCircleCIContextEnvironmentVariable_basic (1.83s)
=== RUN   TestAccCircleCIContextEnvironmentVariable_update
--- PASS: TestAccCircleCIContextEnvironmentVariable_update (3.06s)
=== RUN   TestAccCircleCIContextEnvironmentVariable_import
--- PASS: TestAccCircleCIContextEnvironmentVariable_import (2.13s)
=== RUN   TestAccCircleCIContextEnvironmentVariable_import_name
--- PASS: TestAccCircleCIContextEnvironmentVariable_import_name (2.16s)
=== RUN   TestAccCircleCIContext_basic
--- PASS: TestAccCircleCIContext_basic (1.11s)
=== RUN   TestAccCircleCIContext_update
--- PASS: TestAccCircleCIContext_update (2.48s)
=== RUN   TestAccCircleCIContext_import
--- PASS: TestAccCircleCIContext_import (1.45s)
=== RUN   TestAccCircleCIContext_import_name
--- PASS: TestAccCircleCIContext_import_name (1.61s)
PASS
ok  	github.com/mrolla/terraform-provider-circleci/circleci	17.905s
?   	github.com/mrolla/terraform-provider-circleci/circleci/client	[no test files]
?   	github.com/mrolla/terraform-provider-circleci/circleci/client/rest	[no test files]
```

Will need to confirm the rest of the acceptance tests before considering final details ahead of release.

**Comment by bendrucker** [comment]
All acceptance tests are passing:

```
make testacc
TF_ACC=1 go test ./... -v  -timeout 120m
?   	github.com/mrolla/terraform-provider-circleci	[no test files]
=== RUN   TestAccCircleCIContextDataSource
--- PASS: TestAccCircleCIContextDataSource (1.89s)
=== RUN   TestAccCircleCIContextEnvironmentVariable_basic
--- PASS: TestAccCircleCIContextEnvironmentVariable_basic (1.89s)
=== RUN   TestAccCircleCIContextEnvironmentVariable_update
--- PASS: TestAccCircleCIContextEnvironmentVariable_update (3.03s)
=== RUN   TestAccCircleCIContextEnvironmentVariable_import
--- PASS: TestAccCircleCIContextEnvironmentVariable_import (2.09s)
=== RUN   TestAccCircleCIContextEnvironmentVariable_import_name
--- PASS: TestAccCircleCIContextEnvironmentVariable_import_name (2.01s)
=== RUN   TestAccCircleCIContext_basic
--- PASS: TestAccCircleCIContext_basic (1.02s)
=== RUN   TestAccCircleCIContext_update
--- PASS: TestAccCircleCIContext_update (2.06s)
=== RUN   TestAccCircleCIContext_import
--- PASS: TestAccCircleCIContext_import (1.30s)
=== RUN   TestAccCircleCIContext_import_name
--- PASS: TestAccCircleCIContext_import_name (1.37s)
=== RUN   TestAccCircleCIEnvironmentVariableOrganizationNotSet
--- PASS: TestAccCircleCIEnvironmentVariableOrganizationNotSet (0.01s)
=== RUN   TestAccCircleCIEnvironmentVariableCreateThenUpdateProviderOrg
--- PASS: TestAccCircleCIEnvironmentVariableCreateThenUpdateProviderOrg (4.07s)
=== RUN   TestAccCircleCIEnvironmentVariableCreateThenUpdateResourceOrg
--- PASS: TestAccCircleCIEnvironmentVariableCreateThenUpdateResourceOrg (4.21s)
=== RUN   TestAccCircleCIEnvironmentVariableCreateAlreadyExists
--- PASS: TestAccCircleCIEnvironmentVariableCreateAlreadyExists (2.78s)
=== RUN   TestAccCircleCIEnvironmentVariableImportProviderOrg
--- PASS: TestAccCircleCIEnvironmentVariableImportProviderOrg (2.41s)
=== RUN   TestAccCircleCIEnvironmentVariableImportResourceOrg
--- PASS: TestAccCircleCIEnvironmentVariableImportResourceOrg (2.43s)
=== RUN   TestParseEnvironmentVariableId
--- PASS: TestParseEnvironmentVariableId (0.00s)
=== RUN   TestCircleCIEnvironmentVariableResourceOrgStateUpgradeV0
--- PASS: TestCircleCIEnvironmentVariableResourceOrgStateUpgradeV0 (0.00s)
=== RUN   TestCircleCIEnvironmentVariableProviderOrgStateUpgradeV0
--- PASS: TestCircleCIEnvironmentVariableProviderOrgStateUpgradeV0 (0.00s)
=== RUN   TestValidateEnvironmentVariableName
--- PASS: TestValidateEnvironmentVariableName (0.00s)
PASS
ok  	github.com/mrolla/terraform-provider-circleci/circleci	32.852s
?   	github.com/mrolla/terraform-provider-circleci/circleci/client	[no test files]
?   	github.com/mrolla/terraform-provider-circleci/circleci/client/rest	[no test files]
```

**Comment by bendrucker** [comment]
I've removed the deprecation of provider config for GraphQL. I think this should just be a new major version. Considering we're on 0.x that can be 0.5. The provider API for Terraform users isn't changing at all, but the type of API URL that's accepted has changed. If a user had set the `url` to a custom installation of CircleCI (including Enterprise), this would be breaking.

**Comment by bendrucker** [comment]
@mrolla Do you have time to review this?

File: circleci/resource_circleci_environment_variable_test.go [hunk 1]
```diff
@@ -26,7 +27,7 @@ func TestAccCircleCIEnvironmentVariableOrganizationNotSet(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config:      testAccCircleCIEnvironmentVariableConfigProviderOrg(project, envName, "value-for-the-test"),
-				ExpectError: regexp.MustCompile("organization has not been set for environment variable .*"),
+				ExpectError: regexp.MustCompile("organization"),
```
Comment (mrolla) [inline circleci/resource_circleci_environment_variable_test.go:30]: Is this enough?

File: circleci/resource_circleci_context.go [hunk 2]
```diff
@@ -4,28 +4,20 @@ import (
 	"errors"
 	"fmt"
 	"strings"
-	"time"
-
-	"github.com/CircleCI-Public/circleci-cli/api"
 
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	client "github.com/mrolla/terraform-provider-circleci/circleci/client"
```
Comment (mrolla) [inline circleci/resource_circleci_context.go:9]: nit: would you mind splitting the imports?

File: circleci/resource_circleci_context_environment_variable_test.go [hunk 3]
```diff
@@ -8,10 +8,11 @@ import (
 	"github.com/CircleCI-Public/circleci-cli/api"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+	client "github.com/mrolla/terraform-provider-circleci/circleci/client"
```
Comment (mrolla) [inline circleci/resource_circleci_context_environment_variable_test.go:11]: nit: same thing as above

**Review by mrolla — APPROVED** [review]
Thanks @bendrucker! I only have very minor comments/nits about the structure of the imports (which sometimes is organized in `std`, `3rd` and `1st` and sometimes it's not) and a leftover file in the vendor folder. LGTM :+1: I like how simple it makes things!

I believe this is minor bump worthy, what do you think?

**Comment by bendrucker** [comment]
> I believe this is minor bump worthy, what do you think?

I'd recommend 0.5.

https://semver.org says:

> Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

I prefer to increment `y` for breaking changes or even non-breaking rewrites and `z` for small features and patches. Then after reaching stability (1.0.0) standard semver rules apply.

**Comment by bendrucker** [comment]
I've also added `golangci-lint` here to enforce style conventions and enabled linters that made sense for this project and had no/few violations. [`gci`](https://github.com/daixiang0/gci) can check and handle import ordering automatically, for example.

**Comment by bendrucker** [comment]
Giving you a chance to re-review just in case, otherwise I'll go ahead and merge and release this as 0.5 later this week.

**Comment by mrolla** [comment]
> > I believe this is minor bump worthy, what do you think?
> 
> I'd recommend 0.5.
> 
> https://semver.org says:
> 
> > Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.
> 
> I prefer to increment `y` for breaking changes or even non-breaking rewrites and `z` for small features and patches. Then after reaching stability (1.0.0) standard semver rules apply.

Yep, I agree. That's what I meant with `minor bump worthy`: the minor component should be bumped and release `0.5.0`. I'll review the last 3 commits shortly and re-approve! Thanks!

**Review by mrolla — APPROVED** [review]

**Pull request merged**